src/scheduler/gc_work.rs

```diff
@@ -882,10 +1014,142 @@ impl<E: ProcessEdgesWork, P: Plan<VM = E::VM> + PlanTraceObject<E::VM>> ScanObje
     fn make_another(&self, buffer: Vec<ObjectReference>) -> Self {
         Self::new(self.plan, buffer, self.concurrent, false)
     }
+
+    /// The common code for ScanObjects and PlanScanObjects.
+    /// Which should be different for the different traces
+    fn do_work_common(
+        &self,
+        buffer: &[ObjectReference],
+        worker: &mut GCWorker<<Self::E as ProcessEdgesWork>::VM>,
+        mmtk: &'static MMTK<<Self::E as ProcessEdgesWork>::VM>,
+    ) {
+        let tls = worker.tls;
+
+        // If this is a root packet, the objects in this packet will have not been traced, yet.
+        //
+        // This step conceptually traces the edges from root slots to the objects they point to.
+        // However, VMs that deliver root objects instead of root edges are incapable of updating
+        // root slots.  Like processing an edge, we call `trace_object` on those objects, and
+        // assert the GC doesn't move those objects because we cannot store back to the slots.
+        //
+        // If this is a root packet, the `scanned_root_objects` variable will hold those root
+        // objects which are traced for the first time.
+        let scanned_root_objects = self.roots().then(|| {
+            // We create an instance of E to use its `trace_object` method and its object queue.
+            // let mut process_edges_work = Self::E::new(vec![], false, mmtk);
+            let mut process_edges_work =
+                <PlanProcessEdges<E::VM, P, TRACE_KIND_IMMOVABLE>>::new(vec![], false, mmtk);
```

Comment:
Is this fix to the current `do_work_common()`?
https://github.com/mmtk/mmtk-core/blob/117000feb066edb43e8ebb94c1a465c56fbc771c/src/scheduler/gc_work.rs#L632